sale_timesheet_lock_invoiced/models/account_analytic_line.py

```diff
@@ -0,0 +1,32 @@
+# Copyright 2018 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, models, _
+from odoo.exceptions import ValidationError
+
+
+class AccountAnalyticLine(models.Model):
+
+    _inherit = 'account.analytic.line'
+
+    @api.multi
+    def write(self, values):
+        self._check_locked_lines()
+        return super(AccountAnalyticLine, self).write(values)
+
+    @api.multi
+    def unlink(self):
+        self._check_locked_lines()
+        return super(AccountAnalyticLine, self).unlink()
+
+    @api.multi
+    def _check_locked_lines(self):
+        for rec in self:
+            if rec.task_id and rec.timesheet_invoice_id:
```

Comment:
2 questions:
- why the test on rec.task_id?
- what happens when one deletes a draft invoice linked to a timesheet line?

Reply:
About ```rec.task_id```, it's because the combination of ```task_id``` and ```timesheet_invoice_id``` tells that this line is linked to a sale order line. And we want to lock thoses lines, but indeed, it makes more sense to check ```so_line``` field. I will change it.